src/client/executor.rs

```diff
@@ -105,42 +106,55 @@ impl Client {
         if self.inner.shutdown.executed.load(Ordering::SeqCst) {
             return Err(ErrorKind::Shutdown.into());
         }
-        Box::pin(async {
-            // TODO RUST-9: allow unacknowledged write concerns
-            if !op.is_acknowledged() {
-                return Err(ErrorKind::InvalidArgument {
-                    message: "Unacknowledged write concerns are not supported".to_string(),
-                }
-                .into());
+        // TODO RUST-9: remove this validation
+        if !op.is_acknowledged() {
+            return Err(ErrorKind::InvalidArgument {
+                message: "Unacknowledged write concerns are not supported".to_string(),
             }
-            let session = session.into();
-            if let Some(session) = &session {
-                if !TrackingArc::ptr_eq(&self.inner, &session.client().inner) {
-                    return Err(ErrorKind::InvalidArgument {
-                        message: "the session provided to an operation must be created from the \
-                                  same client as the collection/database"
-                            .into(),
-                    }
-                    .into());
-                }
+            .into());
+        }
+        if let Some(write_concern) = op.write_concern() {
+            write_concern.validate()?;
+        }
 
-                if let Some(SelectionCriteria::ReadPreference(read_preference)) =
-                    op.selection_criteria()
-                {
-                    if session.in_transaction() && read_preference != &ReadPreference::Primary {
-                        return Err(ErrorKind::Transaction {
-                            message: "read preference in a transaction must be primary".into(),
-                        }
-                        .into());
-                    }
+        // Validate the session and update its transaction status if needed.
+        let mut session = session.into();
+        if let Some(ref mut session) = session {
+            if !TrackingArc::ptr_eq(&self.inner, &session.client().inner) {
+                return Err(Error::invalid_argument(
+                    "the session provided to an operation must be created from the same client as \
+                     the collection/database on which the operation is being performed",
+                ));
+            }
+            if op
+                .selection_criteria()
+                .and_then(|sc| sc.as_read_pref())
+                .is_some_and(|rp| rp != &ReadPreference::Primary)
+                && session.in_transaction()
+            {
+                return Err(ErrorKind::Transaction {
+                    message: "read preference in a transaction must be primary".into(),
                 }
+                .into());
             }
-            self.execute_operation_with_retry(op, session).await
-        })
-        .await
+            // If the current transaction has been committed/aborted and it is not being
```

Comment:
I like having this here instead of in `execute_operation_with_retry` - it makes the purpose of that method cleaner - but for my own understanding, was there a functional reason for this change beyond that?

Reply:
nope, this and the reshuffling above were to simplify `execute_operation_with_retry` and `execute_operation_on_connection`